### ISSUE TITLE
Savesettings when user ran with args

### ISSUE THREAD

**Issue by vbjay** [gitextensions/gitextensions#6142]
## Current behaviour

See https://github.com/gitextensions/gitextensions/issues/6060#issuecomment-455914819
We save settings after running a command.

```c#
var commands = new GitUICommands(GetWorkingDir(args));
if (args.Length <= 1)
            {
                commands.StartBrowseDialog();
            }
else
            {
                // if we are here args.Length > 1
                commands.RunCommand(args);
            }

            AppSettings.SaveSettings(); // Right here.  If we run a command we shouldn't save settings.
        }
```

## Expected behaviour

RunCommand should 
- return a value telling us to save settings or not 
- skip saving and in the command running code force a save if desired.

## Steps to reproduce

Based on the comment I linked, I am thinking they ran ```gitext.cmd pull``` and it turned --autostash off for pull dialogs.  
## Environment

- Git Extensions 3.1.0.4945
- Build 231592f2e9e9bce58ebfb02cb124be86e4b4e563 (Dirty)
- Git 2.20.1.windows.1
- Microsoft Windows NT 10.0.17763.0
- .NET Framework 4.7.3260.0
- DPI 96dpi (no scaling)

**Comment by racecrew** [comment]
@vbjay I would like to work on this issue. I have a question/proposal  for better understanding of the issue.

Would it be not enough to check the passed command argument for browse and save the settings? For instance:
```C#
...
var command = args[1];
if (command == "browse")
{
      AppSettings.SaveSettings(); // Right here.  If we run a command we shouldn't save settings.
}
...
```

> return a value telling us to save settings or not

Or do you mean that the code should be indeed refactored/restructured to achieve the mentioned expected behaviour?

**Comment by vbjay** [comment]
It should be a dictionary of the commands and bool, whether the command should or shouldn't update settings. Like browse I would say yes while pull would be no.  Make sure key is compared case insensitive.

**Comment by racecrew** [comment]
```C#
...
SaveSettingsCommandsDict = commands.RunCommand(args);
...
```
You mean, it should go in this direction, don't?

**Comment by vbjay** [comment]
No.
```c#
var dict = new Dictionary<string,bool>()
{["browse"]=false
...}

```

**Comment by racecrew** [comment]
Ok. And where is ot who makes the definition after which command the settings should be saved?

**Comment by vbjay** [comment]
Most will be false. Browse will definitely be true. Once off things like pull should be false since it opens, does it's thing and closes. Reason I say Dictionary is there could be more than browse. So knowing which commands will save or not is first.

**Comment by racecrew** [comment]
Okay, thanks for the explanation... I will do a PR for that.

**Comment by mstv** [comment]
I think a `Dictionary` has several drawbacks:
- need constants for the arguments (duplicate string constants are evil)
- the dictionary will be used only once

So I propose to rather have a `bool saveSettings = false` and set it to `true` in the regarding `case:` for the few affected commands.
In addition or alternatively, `SaveSettings` should send a broadcast to all running GE instances which then reload the settings.
(I have lost modifications to scripts several times!)

**Comment by racecrew** [comment]
>In addition or alternatively, SaveSettings should send a broadcast to all running GE instances which then reload the settings.
(I have lost modifications to scripts several times!)

Where is a broadcast mechanism to complete this task? Is it a part of this issue, or should we create another ticket for that?

**Comment by vbjay** [comment]
Seperate ticket for broadcast.

**Comment by racecrew** [comment]
I am facing in the PR the following issue https://www.codefactor.io/repository/github/gitextensions/gitextensions/pull/8546

Could you please help me? How should I react in such a case? To recode? Is my implementation not okay at all?

Thank you!

**Comment by RussKie** [comment]
> I have lost modifications to scripts several times!

Script handling must be changed like we did for the recent repos. It is on my (loooong) todo list. Since I started using scripts this became a point of annoyance for me as well.
I also want per-repo scripts, but I digress.

**Comment by RussKie** [comment]
First I don't think we should be saving settings at the time the app closes. We had multiple issues with race conditions when users closed several instances of the app at the same time.

All settings must be saved immediately at the point where they change (e.g. when `FormSettings` is closed with [OK], or a toolbar check item is toggled).

**Comment by RussKie** [comment]
First I don't think we should be saving settings at the time the app closes. We had multiple issues with race conditions when users closed several instances of the app at the same time.

All settings must be saved immediately at the point where they change (e.g. when `FormSettings` is closed with [OK], or a toolbar check item is toggled).

@ivangrek has been already cleaning and updating `AppSettings` class. We still need take it further, somewhere where a change to a setting results in save operation. We should support batch changes too (i.e. `BeginUpdate()` - change multiple settings - `EndUpdate()` -> save). `INotifyPropertyChanged` may come in handy here.

Then we should have (if we don't already) file watchers on the settings file, and reload settings whey they change.

**Comment by ivangrek** [comment]
`Move fast and break things. Unless you are breaking stuff, you are not moving fast enough.`

Cash is our minotaur in this maze of settings.

**Comment by ivangrek** [comment]
```
if (!AppSettings.TelemetryEnabled.HasValue)
{
    AppSettings.TelemetryEnabled = MessageBox.Show(null, Strings.TelemetryPermissionMessage,
                                                   Strings.TelemetryPermissionCaption, MessageBoxButtons.YesNo,
                                                   MessageBoxIcon.Question) == DialogResult.Yes;
}
```

Should this show a message for all commands?

I think AppSettings.TelemetryEnabled should not nullable. Instead use other setting to indicate first run of application.

**Comment by ivangrek** [comment]
There will be duplicates of managers.

![image](https://user-images.githubusercontent.com/3169265/96329272-e7aaea00-1053-11eb-824f-88b34ba51644.png)

Black hole?

![image](https://user-images.githubusercontent.com/3169265/96329404-1d040780-1055-11eb-91ca-53c5b8cb3831.png)

**Comment by RussKie** [comment]
@ivangrek apologies, but I'm not sure what you're asking here...

**Comment by ivangrek** [comment]
> @ivangrek apologies, but I'm not sure what you're asking here...

I think this is a dangerous method.

![image](https://user-images.githubusercontent.com/3169265/96337126-535d7900-108d-11eb-9eee-01c2928713ea.png)

No reason to lock here.
Sometimes the settings are saved sometimes not. Perhaps the reason is in this method.

**Comment by RussKie** [comment]
I believe at some point some settings were persisted immediately, and some
weren't. E.g. most changes to the settings in the FormSettigns weren't
persisted until [OK] button was clicked with exception of the scripts (and
something else perhaps?).
IIRC the scripts were refactored and updated, so they only get saved when a
user accepts the changes.
So it is quite possible we can simplify some plumbing.

**Comment by ivangrek** [comment]
@vbjay, perhaps this is the reason and we need to use values for one session.
I think we could remove `AppSettings.SaveSettings();`

https://github.com/gitextensions/gitextensions/pull/8623#issuecomment-733664633

**Comment by ivangrek** [comment]
First commit that adds this https://github.com/gitextensions/gitextensions/commit/adecff1dd1dbba6d20c157fa4fd41325d7718793

**Comment by vbjay** [comment]
We locked to prevent multiple saves concurrently happening in the same
process. Harder to prevent multiple processes from saving at same time.
Think is shutdown an the like.
